### PR TITLE
Tensorrt

### PR DESCRIPTION
**Pull request by fmartinmonier**

### CODE REVIEW

File: darknet_ros/src/YoloObjectDetector.cpp [hunk 1]
```diff
@@ -166,17 +171,64 @@ void YoloObjectDetector::init()
...
+           output_dims_.emplace_back(engine_->getBindingDimensions(i));
+       }
+   }
+   if (input_dims_.empty() || output_dims_.empty())
+   {
+       ROS_ERROR_STREAM("detectionJustYolo::Expect at least one input and one output for network\n");
+   }
+
+  yolo_masks_.push_back(std::vector<int>());
+     yolo_masks_.push_back(std::vector<int>());
```
Comment (marco-tranzatto) [inline darknet_ros/src/YoloObjectDetector.cpp:202]: what are all these "magic numbers" for?

Comment (fmartinmonier) [reply darknet_ros/src/YoloObjectDetector.cpp:202]: I imagine you are talking about the yolo maks and anchors? In which case, here is what I understand of them:
A Yolo network consists of convolutional layers and yolo layers. The convolutional layers behave as normal convolutional layer in a CNN (computing activations of certain features). The yolo layers perform the object detection. Each yolo layers consists of anchor boxes and masks
- Anchors are bounding boxes that are supposed to be (approximately) of the dimension of the objects the user wants to detect. In the yolo layer the image is partitioned in a grid. The intersection over union (IOU) of a portion of a grid (which could potentially contain an artifact) and the anchor boxes is computed. If this value is above a certain threshold, the object is marked as detected. Subsequently, the anchor box is resized to the size of the detected object and becomes the bounding box of the detected object. This avoids having to create a bounding box for each detected object and thus speeds up the network.
- Masks tell the yolo layer which of the anchor boxes is responsible for a high IOU value.

Put simply, these are hyperparameters of the YOLO network. The values are defined in the original version of the network. In this code, they are used in the postprecessing step (after inference in the TensorRT engine).
